Issue #1423: PiecewiseMergeJoin multiple predicates

hawkfish opened the pull request:
Add support for multiple predicates to complex PWMJ.  This provides a 40% performance gain over NestedLoopJoin for range joins:

![image](https://user-images.githubusercontent.com/13156216/154751496-f7f861ae-9c5b-4867-b556-5ccd262bca40.png)

Commit:
Issue #1423: Implement Inequality Joins

Convert PWMJ to merge all the NULL from the condition expressions
into the sort key.

Commit:
Issue #1423: Implement Inequality Joins

Fix small perf regression with one predicate.
Add multi-predicate split into sort key and extra payload.

Commit:
Issue #1423: Implement Inequality Joins

Remove multi-predicate restriction.

Commit:
Issue #1423: Implement Inequality Joins

Merge code review changes.

Commit:
Issue #1423: Implement Inequality Joins

Support multiple predicates for complex joins.

Commit:
Merge branch 'master' into hawkfish-iejoin

Commit:
Merge branch 'master' into hawkfish-iejoin

Commit:
Issue #1423: PWMJ Computed Tails

Compute and compare tail keys on the fly
instead of sorting them. This reduces the run
time of the events 30K benchmark from 24.5s
to 15.9s.

Commit:
Merge branch 'master' into hawkfish-iejoin

Commit:
Merge branch 'master' into hawkfish-iejoin

Commit:
Issue #1423: PWMJ Computed Tails

Fix left block complete vs tail filtering error
that was dropping rows.

Commit:
Issue #1423: PWMJ left Payload

Only extract the left payload once.
This reduces the 30K events benchmark
run time by about 20%.

Commit:
Issue #1423: PWMJ Left Payload

Don't re-extract the left payload for simple joins.

Commit:
Merge branch 'master' into hawkfish-iejoin

Commit:
Issue #1423: PWMJ Right Gather

Only gather the eRHS values that are actually matched.
This improves the events benchmark time by another 15%.

Commit:
Merge branch 'master' into hawkfish-pwmj

Commit:
Issue #1423: PWMJ Right Gather

Reset sorted left payloads.
Use left payloads for outer joins.

Commit:
Issue #1423: PWMJ Right Outer

Fix indexing on right outer joins from vector size testing.

Commit:
Merge branch 'master' into hawkfish-pwmj

Commit:
Merge branch 'master' into hawkfish-pwmj

Commit:
Merge branch 'master' into hawkfish-pwmj

Mytherin commented on test/sql/filter/test_transitive_filters.test:
```diff
@@ -270,6 +270,7 @@ INSERT INTO vals2 SELECT * FROM vals1
 
 query IIII
 SELECT * FROM vals1, vals2 WHERE i>9 AND j<=l AND k>=i AND l<11
+ORDER BY 2 DESC, 4 DESC
```
Could we add some more targeted tests for the multi-predicate piecewise join? While multi-predicate joins do exist in the test suite I think there are not that many. I would like to see e.g.:

* A test involving many (~10) predicates, each filtering out a certain subset of the rows (e.g. `a<z and b<z and c<z and d<z and e<z and ...` where `z=1..100` in random order and `a=10`, `b=20`, `c=30`, etc)
* The same test but with NULLs mixed in as well
* Joins on `BETWEEN` and `NOT BETWEEN` (this should now use the piecewise merge join as well, right?)
* Some more tests with joins on `COMPARE_DISTINCT_FROM` and `COMPARE_NOT_DISTINCT_FROM`, perhaps also checking the plan? I am not sure if these are correctly transformed to piecewise merge joins currently.
* Multi-predicate merge joins with many matches (similar to `test/sql/join/inner/test_unequal_join_duplicates.test` but with a merge join with multiple predicates)

hawkfish replied:
I want to be a bit deliberate about multiple inequalities because IEJoin will hopefully handle two or more of them eventually. So the coverage tests I'm adding will only have one range comparison, but I will add more general tests that will not impact coverage.

hawkfish replied:
`NOT DISTINCT FROM` is not handled by PWMJ (it is an equality condition).

hawkfish replied:
I'm a bit leery of generating too many matches because it is an N^2 algorithm, but I'll interpret that to mean that each side has more than one chunk?

Mytherin reviewed (commented):
Thanks for the PR! Looks good, and great results. Some comments:

Commit:
Merge branch 'master' into hawkfish-pwmj

hawkfish commented on src/execution/operator/join/physical_piecewise_merge_join.cpp:
```diff
@@ -147,6 +177,73 @@ unique_ptr<LocalSinkState> PhysicalPiecewiseMergeJoin::GetLocalSinkState(Executi
+			}
+			pvalidity.EnsureWritable();
+			auto pmask = pvalidity.GetData();
+			if (v.GetVectorType() == VectorType::FLAT_VECTOR) {
```
We need to Orrify because the keys are computed from an input chunk. But there is nothing in that branch that references the vector, so the test just needs to be removed.

Mytherin replied:
Correct me if I'm wrong but the way I'm reading the code below is that it requires the validity mask to hold consecutive values (i.e. it requires the vector to be a flat vector). The `VectorData` struct + selection vector needs to be used to correctly support different vector types.

Perhaps the current code can stay as an optimization for flat vectors (which does not require `Orrify` to be called), and the Orrify can be moved as a back-up strategy for other vector types?

hawkfish replied:
Oh right the masks are tied to the selection vector.

hawkfish commented on src/execution/operator/join/physical_piecewise_merge_join.cpp:
```diff
@@ -721,22 +877,60 @@ OperatorResultType PhysicalPiecewiseMergeJoin::ResolveComplexJoin(ExecutionConte
+					state.rhs_executor.ExecuteExpression(cmp_idx, right);
+
+					if (tail_count < result_count) {
+						left.Slice(*sel, tail_count);
```
Requires three predicates.

hawkfish commented on src/execution/operator/join/physical_piecewise_merge_join.cpp:
```diff
@@ -672,11 +802,37 @@ static idx_t MergeJoinComplexBlocks(BlockMergeInfo &l, BlockMergeInfo &r, const
+	case ExpressionType::COMPARE_LESSTHANOREQUALTO:
+		return VectorOperations::LessThanEquals(left, right, sel, count, true_sel, nullptr);
+	case ExpressionType::COMPARE_GREATERTHANOREQUALTO:
+		return VectorOperations::GreaterThanEquals(left, right, sel, count, true_sel, nullptr);
```
I can add tests here but the goal is to make them uncovered when IEJoin is up and running.

Commit:
Issue #1423: PWMJ Test Coverage

Add test coverage.
Handle non-flat predicate null scanning.

Commit:
Merge branch 'hawkfish-pwmj' of github.com:hawkfish/duckdb into hawkfish-pwmj

Commit:
Issue #1423: IEJoin Events Benchmark

Add a benchmark covering my guess at the IEJoin Q2 data.
Right now, this is being evaluated by PiecewiseMergeJoin.

Mytherin commented:
Thanks!
